File: server/agents/coder_agent.py
import json
import os
from server.Azent.Azent import Agent
from server.tools import get_project_structure


class CoderAgent:
    def __init__(self):
        self.active_sessions = {}

    def get_or_create_agent(self, user_id: str) -> Agent:
        """Get existing agent or create new one for the user"""
        self.next_project_structure = get_project_structure()
        
        if user_id not in self.active_sessions:
            agent = Agent(
                name='coder_agent',
                model="claude-3-5-sonnet-20241022",
                instructions=f'''
                    You are a highly skilled 100x developer which solves and builds tasks in Next.js typescript, Shadcn UI, and Tailwind CSS in a very efficient way.
                    You will be provided with task descriptions and must respond with complete code solutions.
                    You will be provided with the current Nextjs project structure.

                    Nextjs project structure in json format:
                    {json.dumps(self.next_project_structure)}

                    IMPORTANT: You must ALWAYS structure your response in valid JSON format with the following fields:
                    {{
                        "file_name": "Name of the file (e.g., todo.tsx)",
                        "file_path": "Full path to the file (e.g., src/app/todo.tsx)",
                        "code": "The complete code implementation",
                        "description": "Brief explanation of what the code does",
                    }}

                    Do not include any text outside of this JSON structure. Your entire response must be valid JSON that can be parsed directly.
                    ''',
                base_url="https://api.anthropic.com/v1/messages",
                api_key=os.getenv('ANTHROPIC_API_KEY'),
                temperature=0.6,
                session_id=user_id,
                client_type='anthropic'
            )
            self.active_sessions[user_id] = agent
            return agent

        return self.active_sessions[user_id]

    def generate_response(self, user_input: str, user_id: str):
        agent = self.get_or_create_agent(user_id)
        try:
            thread = agent.run(user_input)
            response = [msg for msg in thread if msg['role'] != 'system']

            # You might need to parse the JSON from the text response
            # Depending on your Agent implementation, you might need to add JSON parsing here
            return response
        except Exception as e:
            print(e)
            return [{"role": "assistant", "content": "Something went wrong, please try again later"}]

File: backend/src/agents/coder.agent.ts
import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { RedisCacheService } from 'src/redis/redis.service';
import { BaseAgent, Message } from './base.agent';
import * as fs from 'fs';
import * as path from 'path';

@Injectable()
export class CoderAgent {
  private activeSessions: Map<string, BaseAgent>;

  constructor(
    private readonly configService: ConfigService,
    private readonly redisCacheService: RedisCacheService,
  ) {
    this.activeSessions = new Map();
  }

  private async getOrCreateAgent(userId: string): Promise<BaseAgent> {
    if (!this.activeSessions.has(userId)) {
      // Read base template
      const baseTemplatePath = path.join(
        process.cwd(),
        'src',
        'tools',
        'base-template.json',
      );
      const baseTemplate = JSON.parse(
        fs.readFileSync(baseTemplatePath, 'utf-8'),
      );

      const agent = new BaseAgent(
        this.configService,
        this.redisCacheService,
        'coder_agent',
        'claude-3-5-sonnet-20241022',
        fs
          .readFileSync(
            path.join(process.cwd(), 'src', 'prompts', 'coder-agent.prompt.md'),
            'utf-8',
          )
          .replace('{base_template}', JSON.stringify(baseTemplate)),
        userId,
        0.6,
        [],
        'anthropic',
      );

      this.activeSessions.set(userId, agent);
    }

    return this.activeSessions.get(userId)!;
  }

  async generateResponse(message: string, userId: string): Promise<Message[]> {
    try {
      const agent = await this.getOrCreateAgent(userId);
      const thread = await agent.run(message);
      return thread.filter((msg) => msg.role !== 'system');
    } catch (error) {
      console.error('Error generating response:', error);
      return [
        {
          role: 'assistant',
          content: 'Something went wrong, please try again later',
          type: 'error',
        },
      ];
    }
  }

  clearConversation(userId: string): void {
    this.activeSessions.delete(userId);
  }
}
